src/messenger/messenger.cpp

```diff
@@ -0,0 +1,480 @@
+    return socket->get(zmq::sockopt::routing_id);
+  }
+
+  /**
+   * @return    The next string to be read off the socket.
+   * @warning   Socket must be effectively latched!
+   */
+  static std::string Recv(common::ManagedPointer<zmq::socket_t> socket, zmq::recv_flags flags) {
+    zmq::message_t message;
+    auto received = socket->recv(message, flags);
+    if (!received.has_value()) {
+      throw MESSENGER_EXCEPTION(fmt::format("Unable to receive on socket: {}", ZmqUtil::GetRoutingId(socket)));
+    }
+    return std::string(static_cast<char *>(message.data()), received.value());
+  }
+
+  /**
+   * @return    The next ZmqMessage (identity and payload) read off the socket.
+   * @warning   Socket must be effectively latched!
+   */
+  static messenger::ZmqMessage RecvMsg(common::ManagedPointer<zmq::socket_t> socket) {
+    std::string identity = Recv(socket, zmq::recv_flags::none);
+    std::string delimiter = Recv(socket, zmq::recv_flags::none);
+    std::string payload = Recv(socket, zmq::recv_flags::none);
+
+    return messenger::ZmqMessage::Parse(identity, payload);
+  }
+
+  /**
+   * @return    Send the specified identity over the socket. ROUTER sockets must send this before SendMsgPayload().
+   * @warning   Socket must be effectively latched!
+   */
+  static void SendMsgIdentity(common::ManagedPointer<zmq::socket_t> socket, const std::string &identity) {
+    zmq::message_t identity_msg(identity.data(), identity.size());
+    bool ok = socket->send(identity_msg, zmq::send_flags::sndmore).has_value();
+
+    if (!ok) {
+      throw MESSENGER_EXCEPTION(fmt::format("Unable to send on socket: {}", ZmqUtil::GetRoutingId(socket)));
+    }
+  }
+
+  /**
+   * @return    Send the specified ZmqMessage (delimiter and payload) over the socket.
+   * @warning   Socket must be effectively latched!
+   */
+  static void SendMsgPayload(common::ManagedPointer<zmq::socket_t> socket, const messenger::ZmqMessage &msg) {
+    zmq::message_t delimiter_msg("", 0);
+    zmq::message_t payload_msg(msg.GetRawPayload().data(), msg.GetRawPayload().size());
+    bool ok = true;
+
+    ok = ok && socket->send(delimiter_msg, zmq::send_flags::sndmore).has_value();
+    ok = ok && socket->send(payload_msg, zmq::send_flags::none).has_value();
+
+    if (!ok) {
+      throw MESSENGER_EXCEPTION(fmt::format("Unable to send on socket: {}", ZmqUtil::GetRoutingId(socket)));
+    }
+  }
+};
+
+}  // namespace terrier
+
+namespace terrier::messenger {
+
+ConnectionId::ConnectionId(common::ManagedPointer<Messenger> messenger, const ConnectionDestination &target,
+                           const std::string &identity)
+    : target_name_(target.GetTargetName()) {
+  // Create a new DEALER socket and connect to the server.
+  socket_ = std::make_unique<zmq::socket_t>(*messenger->zmq_ctx_, ZMQ_DEALER);
+  socket_->set(zmq::sockopt::routing_id, identity);
+  socket_->connect(target.GetDestination());
```

Comment:
Should this one be `bind`?

Reply:
No, it should be connect.

You call connect() on sockets that are bind()'d.